arch/arm/src/armv7-a/Toolchain.defs

```diff
@@ -56,21 +56,61 @@ endif
 #
 
 ifeq ($(CONFIG_ARM_THUMB),y)
-ARCHCPUFLAGS += -mthumb
-endif
-
-ifeq ($(CONFIG_ARCH_FPU),y)
-ARCHCPUFLAGS += -mfloat-abi=hard
+  ARCHCPUFLAGS += -mthumb
 endif
 
 ifeq ($(CONFIG_ARCH_CORTEXA5),y)
-ARCHCPUFLAGS += -mcpu=cortex-a5
+  ARCHCPUFLAGS += -mcpu=cortex-a5
 else ifeq ($(CONFIG_ARCH_CORTEXA7),y)
-ARCHCPUFLAGS += -mcpu=cortex-a7
+  ARCHCPUFLAGS += -mcpu=cortex-a7
 else ifeq ($(CONFIG_ARCH_CORTEXA8),y)
-ARCHCPUFLAGS += -mcpu=cortex-a8
+  ARCHCPUFLAGS += -mcpu=cortex-a8
 else ifeq ($(CONFIG_ARCH_CORTEXA9),y)
-ARCHCPUFLAGS += -mcpu=cortex-a9
+  ARCHCPUFLAGS += -mcpu=cortex-a9
+endif
+
+ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
+  ARCHFPUD32FLAGS = -d16
+endif
+
+ifeq ($(CONFIG_ARM_NEON),y)
+  ARCHNEONFLAGS = neon-
+  ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
+    $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
+  endif
+endif
+
+ifeq ($(CONFIG_ARCH_FPU),y)
+  ifeq ($(CONFIG_ARM_FPU_ABI_SOFT),y)
+    ARCHCPUFLAGS += -mfloat-abi=softfp
+  else
+    ARCHCPUFLAGS += -mfloat-abi=hard
+  endif
+
+  ifeq ($(CONFIG_ARCH_CORTEXA7),y)
+    ifneq ($(CONFIG_ARM_NEON),y)
```

Comment:
I think that `vfpv4` should be specified unconditionally. This is according to the table that you provided